doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,979 @@
+    }
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4 ; This instruction...
+      store i32 0, i32* %1     ; ... and this one are useless, they would be deleted if we used an optimization flag.
+      %2 = call i32 @puts(i8* getelementptr inbounds ([13 x i8]* @.str, i32 0, i32 0))
+      ret i32 0
+    }
+
+The two 0 that triggered the debug message from our pass are in the ``store`` and ``ret`` instructions.
+As you can see the compilation to bytecode introduced an expression not present in the source code. This will happen quite often when developing LLVM passes.
+While debugging your future passes you will probably have to read a lot of bytecode so you should familiarize yourself with it.
+Lucky for you it's pretty easy to read (at least compared to asm) and strongly typed (this helps a lot).
+
+
+We've found your replacement
+++++++++++++++++++++++++++++
+
+Now that we can find null literals, we need to be able to replace them.
+We need:
+
+    1. To know the variables reachable from the instruction containing the eligible literal
+    2. To generate the instructions of the arithmetic expression seen earlier
+    3. To insert those expressions back into the code
+    4. (Optional) Generate random prime numbers
+
+Reachable variables
+~~~~~~~~~~~~~~~~~~~
+
+To be sure to have a pool of **reachable** variable during our obfuscation, we are going to register all the variables with integral type we come across while iterating through the block instructions.
+
+We will slightly modify the code to:
+    * add a class member vector storing pointers to the Integer/values of interest. We will empty it at the end of every block.
+    * add a method to check the type of the instruction and store it in the vector if it is eligible.
+    * call the above mentioned method from the main loop.
+
+Our class becomes:
+
+.. code:: C++
+
+    //Add this to your includes
+    #include <vector>
+
+
+    class MyPass : public BasicBlockPass {
+      std::vector<Value *> IntegerVect;
+
+    public:
+
+      static char ID;
+
+      MyPass() : BasicBlockPass(ID) {}
+
+      virtual bool runOnBasicBlock(BasicBlock &BB) override {
+        IntegerVect.clear();
+
+        // Not iterating from the beginning to avoid obfuscation of Phi instructions
+        // parameters
+        for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),
+                                           end = BB.end();
+             I != end; ++I) {
+          Instruction &Inst = *I;
+            for (size_t i{0}; i < Inst.getNumOperands(); ++i) {
+              if (Constant *C = isValidCandidateOperand(Inst.getOperand(i))) {
+                errs() << "I've found one sir!\n";
+              }
+          }
+          registerInteger(Inst);
+        }
+        return false;
+      }
+
+    private:
+      void registerInteger(Value &V) {
+        if (V.getType()->isIntegerTy()) {
+          IntegerVect.emplace_back(&V);
```

Comment:
emplace_back a pointer does not make a difference with push_back, except that the reader will double check the container to be sure because emplace_back is unusual.